File: Data Structures & Algorithms/validate-parentheses/submission-0.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2 != 0:
            return False

        mapping = {
            ')': '(',
            '}': '{',
            ']': '['
        }

        stack = []

        for st in s:
            if st not in mapping:
                stack.append(st)
            elif st in mapping:
                if not stack or stack.pop() != mapping[st]:
                    return False
        return len(stack) == 0

File: Data Structures & Algorithms/decode-string/submission-0.cpp
class Solution {
public:
    string decodeString(string s) {
        stack<int> counts;
        stack<string> strings;
        string curr = "";
        int k = 0;
        
        for (char c : s) {
            if (isdigit(c)) {
                k = k * 10 + (c - '0');
            } else if (c == '[') {
                counts.push(k);
                strings.push(curr);
                k = 0;
                curr = "";
            } else if (c == ']') {
                int repeat = counts.top(); counts.pop();
                string prev = strings.top(); strings.pop();
                string repeated = "";
                for (int i = 0; i < repeat; i++) repeated += curr;
                curr = prev + repeated;
            } else {
                curr += c;
            }
        }
        
        return curr;
    }
};